test/clustertest/BedrockClusterTester.h

```diff
@@ -1,38 +1,222 @@
     // their creation (i.e., node 0 is highest priority and will become leader.
     // You can also specify plugins to load if for some reason you need to override the default configuration.
-    BedrockClusterTester(ClusterSize size, list<string> queries = {}, int threadID = 0, map<string, string> _args = {}, list<string> uniquePorts = {}, string pluginsToLoad = "db,cache,jobs");
-    BedrockClusterTester(int threadID, string pluginsToLoad = "db,cache,jobs");
-    ~BedrockClusterTester();
+    ClusterTester(ClusterSize size, list<string> queries = {}, int threadID = 0, map<string, string> _args = {}, list<string> uniquePorts = {}, string pluginsToLoad = "db,cache,jobs");
+    ClusterTester(int threadID, string pluginsToLoad = "db,cache,jobs");
+    ClusterTester(const string& pluginString = "db,cache,jobs");
+    ~ClusterTester();
 
-    // Returns the bedrock tester at the given index in the cluster.
-    BedrockTester* getBedrockTester(size_t index);
+    // Returns the tester at the given index in the cluster.
+    T& getTester(size_t index);
+
+    // Legacy name for the above. TODO: Remove when not used by auth.
+    T* getBedrockTester(size_t index);
 
     // Starts a given node, given the same arguments given by the constructor.
-    string startNode(size_t nodeIndex);
+    string startNode(size_t index);
 
     // Same as above but don't wait for the command port to be ready.
-    string startNodeDontWait(size_t nodeIndex);
+    string startNodeDontWait(size_t index);
 
     // Stops a given node.
-    void stopNode(size_t nodeIndex);
+    void stopNode(size_t index);
 
   private:
 
     // The number of nodes in the cluster.
     int _size;
 
     // A list of all our testers that make up our cluster.
-    vector<BedrockTester> _cluster;
+    list<T> _cluster;
 };
+
+typedef ClusterTester<BedrockTester> BedrockClusterTester;
+
+template <typename T>
+ClusterTester<T>::ClusterTester(const string& pluginString) : ClusterTester<T>(
+    ClusterSize::THREE_NODE_CLUSTER,
+    {},
+    0,
+    {},
+    {},
+    pluginString
+    )
+{}
+
+template <typename T>
+ClusterTester<T>::ClusterTester(int threadID, string pluginsToLoad)
+  : ClusterTester<T>(ClusterSize::THREE_NODE_CLUSTER, {},
+                            threadID, {}, {}, pluginsToLoad, nullptr) {}
+
+template <typename T>
+ClusterTester<T>::ClusterTester(ClusterSize size,
+                                list<string> queries,
+                                int threadID,
+                                map<string, string> _args,
+                                list<string> uniquePorts,
+                                string pluginsToLoad)
+: _size((int)size)
+{
+    // We need three ports for each node.
+    vector<vector<uint16_t>> ports((int)size);
+    for (size_t i = 0; i < (size_t)size; i++) {
+        const uint16_t serverPort = BedrockTester::ports.getPort();
```

Comment:
this constructor takes a generic template but requires BedrockTester functions? that's kind of odd

Reply:
Realistically, this is only going to work for subclasses of `BedrockTester` or very similar classes that imitate `BedrockTester`. It's not like you'll be able to make a `ClusterTester<string>`.

We can create some other class that's effectively `SingletonTCPPortAllocator` but I don't think it's going to ever be used for anything that's not a `BedrockTester` anyway.